File: server/contracts.py
from __future__ import annotations

import unicodedata
from datetime import datetime, timezone
from re import compile as compile_regex
from typing import Annotated, Literal
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    StringConstraints,
    field_validator,
    model_validator,
)

from .errors import ErrorCode


ShortText = Annotated[str, StringConstraints(min_length=1, max_length=240)]
InstallationToken = Annotated[str, StringConstraints(min_length=1, max_length=2_048)]
AccountToken = Annotated[str, StringConstraints(min_length=1, max_length=2_048)]
RevenueCatInstallationId = Annotated[
    str,
    StringConstraints(
        min_length=18,
        max_length=160,
        pattern=r"^rai_installation_[A-Za-z0-9_-]+$",
    ),
]
RevenueCatAccountId = Annotated[
    str,
    StringConstraints(
        min_length=13,
        max_length=160,
        pattern=r"^rai_account_[A-Za-z0-9_-]+$",
    ),
]
FeedbackText = Annotated[str, StringConstraints(min_length=1, max_length=600)]
SimulatedRecruiterComment = Annotated[
    str,
    StringConstraints(
        min_length=1,
        max_length=800,
        pattern=r"^Simulated AI recruiter feedback:",
    ),
]
CANONICAL_UUID_V2 = compile_regex(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$"
)
CANONICAL_UTC_SECOND = compile_regex(
    r"^[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}Z$"
)


def _normalized_keyword(value: str) -> str:
    return unicodedata.normalize("NFKC", value).strip().casefold()


def _parse_canonical_utc_second(value: object, *, name: str) -> datetime:
    if not isinstance(value, str) or CANONICAL_UTC_SECOND.fullmatch(value) is None:
        raise ValueError(f"{name} must be a canonical whole-second UTC timestamp")
    try:
        parsed = datetime.strptime(value, "%Y-%m-%dT%H:%M:%SZ")
    except ValueError:
        raise ValueError(
            f"{name} must be a canonical whole-second UTC timestamp"
        ) from None
    return parsed.replace(tzinfo=timezone.utc)


class StrictContract(BaseModel):
    model_config = ConfigDict(extra="forbid", strict=True)


class ScoreComponentsV1(StrictContract):
    model_config = ConfigDict(
        extra="forbid",
        strict=True,
        json_schema_extra={
            "allOf": [
                {
                    "if": {
                        "properties": {"keywords": {"type": "null"}},
                        "required": ["keywords"],
                    },
                    "then": {
                        "properties": {
                            "structure": {"maximum": 30},
                            "impact": {"maximum": 40},
                            "readability": {"maximum": 30},
                            "keywords": {"type": "null"},
                        }
                    },
                    "else": {
                        "properties": {
                            "structure": {"maximum": 25},
                            "impact": {"maximum": 30},
                            "readability": {"maximum": 20},
                            "keywords": {
                                "type": "integer",
                                "minimum": 0,
                                "maximum": 25,
                            },
                        }
                    },
                }
            ]
        },
    )
    structure: int = Field(ge=0, le=30)
    impact: int = Field(ge=0, le=40)
    readability: int = Field(ge=0, le=30)
    keywords: int | None = Field(ge=0, le=25)

    @model_validator(mode="after")
    def enforces_scoring_branch_maxima(self) -> ScoreComponentsV1:
        maxima = (
            {"structure": 30, "impact": 40, "readability": 30}
            if self.keywords is None
            else {"structure": 25, "impact": 30, "readability": 20}
        )
        values = {
            "structure": self.structure,
            "impact": self.impact,
            "readability": self.readability,
        }
        for field, maximum in maxima.items():
            if values[field] > maximum:
                raise ValueError(f"{field} exceeds its scoring-branch maximum")
        return self


class ScoreV1(StrictContract):
    scoreVersion: Literal["resume-readiness-v1"]
    readinessScore: int = Field(ge=0, le=100)
    label: Literal["Needs work", "Developing", "Good", "Strong"]
    components: ScoreComponentsV1
    explanations: list[ShortText] = Field(max_length=12)

    @model_validator(mode="after")
    def has_consistent_score(self) -> ScoreV1:
        component_total = sum(
            value for value in self.components.model_dump().values() if value is not None
        )
        if self.readinessScore != component_total:
            raise ValueError("readinessScore must equal the component total")
        expected_label = (
            "Needs work"
            if self.readinessScore < 50
            else "Developing"
            if self.readinessScore < 70
            else "Good"
            if self.readinessScore < 85
            else "Strong"
        )
        if self.label != expected_label:
            raise ValueError("label must match readinessScore")
        return self


class FeedbackV1(StrictContract):
    matchedKeywords: list[FeedbackText] = Field(max_length=20)
    missingKeywords: list[FeedbackText] = Field(max_length=20)
    strengths: list[FeedbackText] = Field(min_length=1, max_length=12)
    improvements: list[FeedbackText] = Field(min_length=1, max_length=12)
    powerBullets: list[FeedbackText] = Field(max_length=10)
    summary: Annotated[str, StringConstraints(min_length=1, max_length=500)]
    simulatedRecruiterComment: SimulatedRecruiterComment

    @model_validator(mode="after")
    def has_consistent_keyword_lists(self) -> FeedbackV1:
        matched = {_normalized_keyword(value) for value in self.matchedKeywords}
        missing = {_normalized_keyword(value) for value in self.missingKeywords}
        if matched & missing:
            raise ValueError("matchedKeywords and missingKeywords must not overlap")
        return self


class AnalysisResponseV1(StrictContract):
    schemaVersion: Literal[1]
    analysisId: UUID
    sourceType: Literal["pdf", "text", "vision_text"]
    score: ScoreV1
    feedback: FeedbackV1

    @field_validator("analysisId", mode="before")
    @classmethod
    def parses_json_uuid(cls, value: object) -> object:
        return UUID(value) if isinstance(value, str) else value


AiStatusV2 = Literal[
    "complete",
    "not_requested",
    "quota_exhausted",
    "plan_verification_unavailable",
    "temporarily_unavailable",
    "timeout",
    "invalid_provider_response",
]


class AiAllowanceV2(StrictContract):
    used: int = Field(ge=0, le=100)
    limit: Literal[3, 100]
    resetsAt: datetime

    @field_validator("resetsAt", mode="before")
    @classmethod
    def parses_canonical_utc_second(cls, value: object) -> object:
        return _parse_canonical_utc_second(value, name="resetsAt")

    @model_validator(mode="after")
    def used_does_not_exceed_limit(self) -> AiAllowanceV2:
        if self.used > self.limit:
            raise ValueError("used exceeds allowance limit")
        return self


class AiResultV2(StrictContract):
    status: AiStatusV2
    feedback: FeedbackV1 | None
    allowance: AiAllowanceV2 | None

    @model_validator(mode="after")
    def enforces_status_payload(self) -> AiResultV2:
        if self.status == "complete" and (
            self.feedback is None or self.allowance is None
        ):
            raise ValueError("complete AI results require feedback and allowance")
        if self.status != "complete" and self.feedback is not None:
            raise ValueError("feedback is present only for complete AI results")
        return self


class AnalysisResponseV2(StrictContract):
    schemaVersion: Literal[2]
    analysisId: UUID
    sourceType: Literal["reviewed_text", "pdf"]
    score: ScoreV1
    ai: AiResultV2

    @field_validator("analysisId", mode="before")
    @classmethod
    def parses_canonical_json_uuid(cls, value: object) -> object:
        if not isinstance(value, str) or CANONICAL_UUID_V2.fullmatch(value) is None:
            raise ValueError("analysisId must use canonical lowercase UUID spelling")
        return UUID(value)


class InstallationResponseV1(StrictContract):
    schemaVersion: Literal[1]
    installationToken: InstallationToken


class PlanSnapshotV2(StrictContract):
    schemaVersion: Literal[2]
    plan: Literal["free", "pro"]
    verifiedUntil: datetime
    entitlementExpiresAt: datetime | None
    allowance: AiAllowanceV2

    @field_validator("verifiedUntil", "entitlementExpiresAt", mode="before")
    @classmethod
    def parses_canonical_plan_timestamp(cls, value: object, info) -> object:
        if value is None and info.field_name == "entitlementExpiresAt":
            return None
        return _parse_canonical_utc_second(value, name=info.field_name)

    @model_validator(mode="after")
    def enforces_plan_payload(self) -> PlanSnapshotV2:
        if self.plan == "free":
            if self.entitlementExpiresAt is not None or self.allowance.limit != 3:
                raise ValueError("free plan payload is invalid")
        else:
            if (
                self.entitlementExpiresAt is None
                or self.verifiedUntil > self.entitlementExpiresAt
                or self.allowance.limit != 100
            ):
                raise ValueError("pro plan payload is invalid")
        return self


class EntitlementSyncRequestV2(StrictContract):
    pass


class AppleIdentityRequestV2(StrictContract):
    identityToken: Annotated[str, StringConstraints(min_length=1, max_length=8_192)] = Field(
        repr=False
    )
    nonce: Annotated[str, StringConstraints(min_length=8, max_length=256)] = Field(
        repr=False
    )

    @field_validator("identityToken", "nonce")
    @classmethod
    def rejects_control_characters(cls, value: str) -> str:
        if any(ord(character) < 32 or ord(character) == 127 for character in value):
            raise ValueError("sensitive identity field is invalid")
        return value


class AppleIdentityResponseV2(StrictContract):
    schemaVersion: Literal[2]
    accountToken: AccountToken = Field(repr=False)
    expiresAt: datetime
    revenueCatAppUserId: RevenueCatAccountId = Field(repr=False)

    @field_validator("expiresAt", mode="before")
    @classmethod
    def parses_canonical_expiry(cls, value: object) -> object:
        return _parse_canonical_utc_second(value, name="expiresAt")


class InstallationResponseV2(StrictContract):
    schemaVersion: Literal[2]
    installationToken: InstallationToken = Field(repr=False)
    revenueCatAppUserId: RevenueCatInstallationId = Field(repr=False)


class PublicErrorV1(StrictContract):
    schemaVersion: Literal[1]
    code: ErrorCode
    message: Annotated[str, StringConstraints(min_length=1, max_length=240)]
    requestId: UUID
    retryable: bool

    @field_validator("requestId", mode="before")
    @classmethod
    def parses_json_uuid(cls, value: object) -> object:
        return UUID(value) if isinstance(value, str) else value

    @field_validator("code", mode="before")
    @classmethod
    def parses_json_error_code(cls, value: object) -> object:
        return ErrorCode(value) if isinstance(value, str) else value

File: mobile/src/domain/contracts.ts
import { z } from 'zod';

import {
  casefoldCharacterUnicode15,
  hasUnicode15Data,
  isPythonStripCharacter,
  normalizeNfkcUnicode15,
} from './generated/unicode15';
import { codePointLength } from './limits';

const ERROR_CODES = [
  'invalid_request',
  'invalid_installation',
  'rate_limited',
  'request_in_progress',
  'unsupported_file',
  'file_too_large',
  'pdf_too_many_pages',
  'pdf_encrypted',
  'pdf_invalid',
  'pdf_timeout',
  'scan_required',
  'resume_too_long',
  'scoring_input_limit',
  'ai_timeout',
  'ai_unavailable',
  'invalid_ai_response',
  'service_misconfigured',
  'service_unavailable',
] as const;

const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;

function codePointText(minimum: number, maximum: number) {
  return z.string().superRefine((value, context) => {
    const length = codePointLength(value);
    if (length < minimum || length > maximum) {
      context.addIssue({
        code: 'custom',
        message: `Text must contain ${minimum} to ${maximum} Unicode code points.`,
      });
    }
  });
}

function labelFor(readinessScore: number): 'Needs work' | 'Developing' | 'Good' | 'Strong' {
  if (readinessScore < 50) return 'Needs work';
  if (readinessScore < 70) return 'Developing';
  if (readinessScore < 85) return 'Good';
  return 'Strong';
}

export function normalizeKeywordForService(value: string): string {
  if (!hasUnicode15Data()) throw new Error('Unicode service data is unavailable.');
  const normalized = normalizeNfkcUnicode15(value);
  if (normalized === null) throw new Error('Unicode service data is unavailable.');
  const characters = Array.from(normalized);
  let start = 0;
  let end = characters.length;
  while (start < end && isPythonStripCharacter(characters[start])) start += 1;
  while (end > start && isPythonStripCharacter(characters[end - 1])) end -= 1;
  let folded = '';
  for (const character of characters.slice(start, end)) {
    const mapped = casefoldCharacterUnicode15(character);
    if (mapped === null) throw new Error('Unicode service data is unavailable.');
    folded += mapped;
  }
  return folded;
}

export const ScoreComponentsSchema = z
  .object({
    structure: z.number().int().min(0).max(30),
    impact: z.number().int().min(0).max(40),
    readability: z.number().int().min(0).max(30),
    keywords: z.number().int().min(0).max(25).nullable(),
  })
  .strict()
  .superRefine((components, context) => {
    const maxima = components.keywords === null
      ? { structure: 30, impact: 40, readability: 30 }
      : { structure: 25, impact: 30, readability: 20 };
    for (const key of ['structure', 'impact', 'readability'] as const) {
      if (components[key] > maxima[key]) {
        context.addIssue({
          code: 'custom',
          message: `${key} exceeds its scoring-branch maximum`,
          path: [key],
        });
      }
    }
  });

export const ScoreSchema = z
  .object({
    scoreVersion: z.literal('resume-readiness-v1'),
    readinessScore: z.number().int().min(0).max(100),
    label: z.enum(['Needs work', 'Developing', 'Good', 'Strong']),
    components: ScoreComponentsSchema,
    explanations: z.array(codePointText(1, 240)).max(12),
  })
  .strict()
  .superRefine((score, context) => {
    const total =
      score.components.structure +
      score.components.impact +
      score.components.readability +
      (score.components.keywords ?? 0);
    if (total !== score.readinessScore) {
      context.addIssue({ code: 'custom', message: 'score component total mismatch' });
    }
    if (labelFor(score.readinessScore) !== score.label) {
      context.addIssue({ code: 'custom', message: 'score label mismatch' });
    }
  });

export const FeedbackSchema = z
  .object({
    matchedKeywords: z.array(codePointText(1, 600)).max(20),
    missingKeywords: z.array(codePointText(1, 600)).max(20),
    strengths: z.array(codePointText(1, 600)).min(1).max(12),
    improvements: z.array(codePointText(1, 600)).min(1).max(12),
    powerBullets: z.array(codePointText(1, 600)).max(10),
    summary: codePointText(1, 500),
    simulatedRecruiterComment: codePointText(1, 800).startsWith(
      'Simulated AI recruiter feedback:',
    ),
  })
  .strict()
  .superRefine((feedback, context) => {
    try {
      const matched = new Set(feedback.matchedKeywords.map(normalizeKeywordForService));
      if (feedback.missingKeywords.some((item) => matched.has(normalizeKeywordForService(item)))) {
        context.addIssue({ code: 'custom', message: 'keyword lists overlap' });
      }
    } catch {
      context.addIssue({ code: 'custom', message: 'Unicode service data unavailable' });
    }
  });

export const AnalysisResponseSchema = z
  .object({
    schemaVersion: z.literal(1),
    analysisId: z.string().regex(UUID),
    sourceType: z.enum(['pdf', 'text', 'vision_text']),
    score: ScoreSchema,
    feedback: FeedbackSchema,
  })
  .strict();

export const AiAllowanceSchema = z
  .object({
    used: z.number().int().min(0).max(100),
    limit: z.union([z.literal(3), z.literal(100)]),
    resetsAt: z.string().datetime({ offset: false, precision: 0 }),
  })
  .strict()
  .refine(value => value.used <= value.limit, {
    message: 'used exceeds allowance limit',
    path: ['used'],
  });

const AiUnavailableSchema = z
  .object({
    status: z.enum([
      'not_requested',
      'quota_exhausted',
      'plan_verification_unavailable',
      'temporarily_unavailable',
      'timeout',
      'invalid_provider_response',
    ]),
    feedback: z.null(),
    allowance: AiAllowanceSchema.nullable(),
  })
  .strict();

export const AiResultSchema = z.discriminatedUnion('status', [
  z
    .object({
      status: z.literal('complete'),
      feedback: FeedbackSchema,
      allowance: AiAllowanceSchema,
    })
    .strict(),
  AiUnavailableSchema,
]);

export const AnalysisResponseV2Schema = z
  .object({
    schemaVersion: z.literal(2),
    analysisId: z.string().regex(UUID),
    sourceType: z.enum(['reviewed_text', 'pdf']),
    score: ScoreSchema,
    ai: AiResultSchema,
  })
  .strict();

export const InstallationResponseSchema = z
  .object({
    schemaVersion: z.literal(1),
    installationToken: codePointText(1, 2048),
  })
  .strict();

export const PublicErrorSchema = z
  .object({
    schemaVersion: z.literal(1),
    code: z.enum(ERROR_CODES),
    message: codePointText(1, 240),
    requestId: z.string().regex(UUID),
    retryable: z.boolean(),
  })
  .strict();

const AnalysisRequestContextSchema = z.object({ hasJobDescription: z.boolean() }).strict();

export type AnalysisResponse = z.infer<typeof AnalysisResponseSchema>;
export type AnalysisResponseV2 = z.infer<typeof AnalysisResponseV2Schema>;
export type InstallationResponse = z.infer<typeof InstallationResponseSchema>;
export type PublicError = z.infer<typeof PublicErrorSchema>;
export type AnalysisRequestContext = Readonly<z.infer<typeof AnalysisRequestContextSchema>>;

export function parseAnalysisResponse(
  value: unknown,
  context: AnalysisRequestContext,
): AnalysisResponse {
  const parsedContext = AnalysisRequestContextSchema.parse(context);
  const response = AnalysisResponseSchema.parse(value);
  const keywordsAreNull = response.score.components.keywords === null;
  if (keywordsAreNull === parsedContext.hasJobDescription) {
    throw new z.ZodError([
      {
        code: 'custom',
        message: 'job-description keyword component mismatch',
        path: ['score', 'components', 'keywords'],
      },
    ]);
  }
  return response;
}
